x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportGetShutdownStatusAction.java

```diff
@@ -111,6 +150,83 @@ protected void masterOperation(
         listener.onResponse(response);
     }
 
+    // pkg-private for testing
+    static ShutdownShardMigrationStatus shardMigrationStatus(
+        ClusterState currentState,
+        String nodeId,
+        SingleNodeShutdownMetadata.Type shutdownType,
+        AllocationDeciders allocationDeciders,
+        ClusterInfoService clusterInfoService,
+        SnapshotsInfoService snapshotsInfoService,
+        AllocationService allocationService
+    ) {
+        // Only REMOVE-type shutdowns will try to move shards, so RESTART-type shutdowns should immediately complete
+        if (SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownType)) {
+            return new ShutdownShardMigrationStatus(SingleNodeShutdownMetadata.Status.COMPLETE, 0);
+        }
+
+        // First, check if there are any shards currently on this node, and if there are any relocating shards
+        int currentShardsOnNode = currentState.getRoutingNodes().node(nodeId).numberOfShardsWithState(ShardRoutingState.STARTED);
+        int currentlyRelocatingShards = currentState.getRoutingNodes().node(nodeId).numberOfShardsWithState(ShardRoutingState.RELOCATING);
+        int totalRemainingShards = currentlyRelocatingShards + currentShardsOnNode;
```

Comment:
I think this is missing dealing with `INITIALIZING` shards, which may be assigned but are part of the total? Should those be considered as shards currently on the node that need to be migrated?

Reply:
Very good question. Since the allocation decider should prevent any new shards from being allocated to the node, the only `INITIALIZING` shards should be ones which were already assigned to the node before the shutdown was registered, but there still might be some. 

In what circumstances would shards stay `INITIALIZING` for a long time? I'm not sure if snapshot restoration shows up as that, or another status - do you know off the top of your head?

`INITIALIZING` shards can't be relocated per my understanding, so we'd have to wait until they move to `STARTED`... and then would they relocate? Or would we have to trigger a reroute again? I think I'm going to have to ask the distributed team about some of this.

Reply:
> In what circumstances would shards stay `INITIALIZING` for a long time? I'm not sure if snapshot restoration shows up as that, or another status - do you know off the top of your head?

Initial recovery during startup as well as opening a closed index moves shards to initializing I *think*, and I think if a user set the `check_on_startup` setting it can take quite a while. I don't know about snapshot restoration unfortunately.

> `INITIALIZING` shards can't be relocated per my understanding, so we'd have to wait until they move to `STARTED`... and then would they relocate? Or would we have to trigger a reroute again?

I believe they cannot be relocated during `INITIALIZING` state (it would make sense you can't relocate a shard recovering from local disk), and I *think* they will be subject to a relocation/reroute once they've reached the `STARTED` state, but don't hold me to that :)

Reply:
Henning told me elsewhere that (among other things):
> If we have cases of shards getting "stuck" in INITIALIZING, I would call it a bug, but we have seen cases of slowness (which is possibly also a bug).

So we *shouldn't* get cases of shards being stuck in `INITIALIZING` indefinitely, but there may be some where they are very slow to move out of that state. So this now counts initializing shards towards the shards remaining count, and if there are _only_ initializing shards left, reports `IN_PROGRESS` with a `reason` explaining that there are only `INITIALIZING` shards left. That should help surface unexpected conditions more quickly. I'm not sure we can get much more detail there easily, though.